### PR TITLE
BattleContext class

### LINKED ISSUE #15: BattleContext Class
## Background

A battle context describes the POV of a given Pokemon on the battlefield. It provides enough information for a user to make an informed decision about what turn to make next.

Fields
- `Battle`: The state of the battlefield
- `Pokemon`: The pokemon that this context belongs to.
- `Team`: The team ID of the `Pokemon`
- `Targets`: All pokemon on the battlefield
- `Allies`: Ally targets in relation to the `Pokemon`
- `Opponents`: Enemy targets in relation to the `Pokemon`

## Objectives
- Create a class called `BattleContext` in `pkmntypes.py`
- Create a function/constructor to parse the json into an instance of `BattleContext`
    - In Python, when JSON is parsed, it gets turned into a `dict`. So the function should take a `dict` as input, not a string.
    - Parsing the pokemon objects is not necessary for this issue, as that should be taken care of by #16 .

<details>
<summary>Example JSON</summary>

```json
{
    "Battle": {
        "Weather": 0,
        "ShiftSet": false,
        "State": 1
    },
    "Pokemon": {
        "Name": "Pichu",
        "NatDex": 172,
        "Level": 79,
        "Ability": 98,
        "TotalExperience": 493039,
        "Gender": 0,
        "IVs": [
            1,
            2,
            4,
            0,
            2,
            1
        ],
        "EVs": [
            18,
            17,
            22,
            32,
            21,
            41
        ],
        "Nature": 8,
        "Stats": [
            124,
            72,
            35,
            66,
            65,
            108
        ],
        "StatModifiers": [
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0
        ],
        "StatusEffects": 0,
        "CurrentHP": 124,
        "HeldItem": {
            "Id": 0,
            "Name": "",
            "Category": 0,
            "FlingPower": 0,
            "FlingEffect": 0,
            "Flags": 0
        },
        "Moves": [
            {
                "Id": 433,
                "CurrentPP": 0,
                "MaxPP": 0,
                "Name": "Trick Room",
                "Type": 8192,
                "Category": 0,
                "Targets": 12,
                "Priority": -7,
                "Power": 0,
                "Accuracy": 0,
                "InitialMaxPP": 0
            },
            {
                "Id": 225,
                "CurrentPP": 0,
                "MaxPP": 0,
                "Name": "Dragon Breath",
                "Type": 32768,
                "Category": 2,
                "Targets": 10,
                "Priority": 0,
                "Power": 60,
                "Accuracy": 100,
                "InitialMaxPP": 0
            },
            {
                "Id": 232,
                "CurrentPP": 0,
                "MaxPP": 0,
                "Name": "Metal Claw",
                "Type": 256,
                "Category": 1,
                "Targets": 10,
                "Priority": 0,
                "Power": 50,
                "Accuracy": 95,
                "InitialMaxPP": 0
            },
            {
                "Id": 59,
                "CurrentPP": 0,
                "MaxPP": 0,
                "Name": "Blizzard",
                "Type": 16384,
                "Category": 2,
                "Targets": 11,
                "Priority": 0,
                "Power": 110,
                "Accuracy": 70,
                "InitialMaxPP": 0
            }
        ],
        "Friendship": 0,
        "OriginalTrainerID": 0,
        "Type": 4096
    },
    "Team": 0,
    "Allies": [
        {
            "Party": 0,
            "Slot": 0,
            "Team": 0,
            "Pokemon": {
                "Name": "Pichu",
                "NatDex": 172,
                "Level": 79,
                "Ability": 98,
                "TotalExperience": 493039,
                "Gender": 0,
                "IVs": [
                    1,
                    2,
                    4,
                    0,
                    2,
                    1
                ],
                "EVs": [
                    18,
                    17,
                    22,
                    32,
                    21,
                    41
                ],
                "Nature": 8,
                "Stats": [
                    124,
                    72,
                    35,
                    66,
                    65,
                    108
                ],
                "StatModifiers": [
                    0,
                    0,
                    0,
                    0,
                    0,
                    0,
                    0,
                    0,
                    0
                ],
                "StatusEffects": 0,
                "CurrentHP": 124,
                "HeldItem": {
                    "Id": 0,
                    "Name": "",
                    "Category": 0,
                    "FlingPower": 0,
                    "FlingEffect": 0,
                    "Flags": 0
                },
                "Moves": [
                    {
                        "Id": 433,
                        "CurrentPP": 0,
                        "MaxPP": 0,
                        "Name": "Trick Room",
                        "Type": 8192,
                        "Category": 0,
                        "Targets": 12,
                        "Priority": -7,
                        "Power": 0,
                        "Accuracy": 0,
                        "InitialMaxPP": 0
                    },
                    {
                        "Id": 225,
                        "CurrentPP": 0,
                        "MaxPP": 0,
                        "Name": "Dragon Breath",
                        "Type": 32768,
                        "Category": 2,
                        "Targets": 10,
                        "Priority": 0,
                        "Power": 60,
                        "Accuracy": 100,
                        "InitialMaxPP": 0
                    },
                    {
                        "Id": 232,
                        "CurrentPP": 0,
                        "MaxPP": 0,
                        "Name": "Metal Claw",
                        "Type": 256,
                        "Category": 1,
                        "Targets": 10,
                        "Priority": 0,
                        "Power": 50,
                        "Accuracy": 95,
                        "InitialMaxPP": 0
                    },
                    {
                        "Id": 59,
                        "CurrentPP": 0,
                        "MaxPP": 0,
                        "Name": "Blizzard",
                        "Type": 16384,
                        "Category": 2,
                        "Targets": 11,
                        "Priority": 0,
                        "Power": 110,
                        "Accuracy": 70,
                        "InitialMaxPP": 0
                    }
                ],
                "Friendship": 0,
                "OriginalTrainerID": 0,
                "Type": 4096
            }
        }
    ],
    "Opponents": [
        {
            "Party": 0,
            "Slot": 0,
            "Team": 1,
            "Pokemon": {
                "Name": "Mew",
                "NatDex": 151,
                "Level": 6,
                "Ability": 108,
                "TotalExperience": 179,
                "Gender": 0,
                "IVs": [
                    3,
                    1,
                    1,
                    4,
                    0,
                    3
                ],
                "EVs": [
                    7,
                    24,
                    41,
                    11,
                    3,
                    20
                ],
                "Nature": 8,
                "Stats": [
                    28,
                    17,
                    17,
                    17,
                    17,
                    17
                ],
                "StatModifiers": [
                    0,
                    0,
                    0,
                    0,
                    0,
                    0,
                    0,
                    0,
                    0
                ],
                "StatusEffects": 0,
                "CurrentHP": 28,
                "HeldItem": {
                    "Id": 0,
                    "Name": "",
                    "Category": 0,
                    "FlingPower": 0,
                    "FlingEffect": 0,
                    "Flags": 0
                },
                "Moves": [
                    {
                        "Id": 61,
                        "CurrentPP": 0,
                        "MaxPP": 0,
                        "Name": "Bubble Beam",
                        "Type": 1024,
                        "Category": 2,
                        "Targets": 10,
                        "Priority": 0,
                        "Power": 65,
                        "Accuracy": 100,
                        "InitialMaxPP": 0
                    },
                    {
                        "Id": 457,
                        "CurrentPP": 0,
                        "MaxPP": 0,
                        "Name": "Head Smash",
                        "Type": 32,
                        "Category": 1,
                        "Targets": 10,
                        "Priority": 0,
                        "Power": 150,
                        "Accuracy": 80,
                        "InitialMaxPP": 0
                    },
                    {
                        "Id": 434,
                        "CurrentPP": 0,
                        "MaxPP": 0,
                        "Name": "Draco Meteor",
                        "Type": 32768,
                        "Category": 2,
                        "Targets": 10,
                        "Priority": 0,
                        "Power": 130,
                        "Accuracy": 90,
                        "InitialMaxPP": 0
                    },
                    {
                        "Id": 359,
                        "CurrentPP": 0,
                        "MaxPP": 0,
                        "Name": "Hammer Arm",
                        "Type": 2,
                        "Category": 1,
                        "Targets": 10,
                        "Priority": 0,
                        "Power": 100,
                        "Accuracy": 90,
                        "InitialMaxPP": 0
                    }
                ],
                "Friendship": 0,
                "OriginalTrainerID": 0,
                "Type": 8192
            }
        }
    ],
    "Targets": [
        {
            "Party": 0,
            "Slot": 0,
            "Team": 0,
            "Pokemon": {
                "Name": "Pichu",
                "NatDex": 172,
                "Level": 79,
                "Ability": 98,
                "TotalExperience": 493039,
                "Gender": 0,
                "IVs": [
                    1,
                    2,
                    4,
                    0,
                    2,
                    1
                ],
                "EVs": [
                    18,
                    17,
                    22,
                    32,
                    21,
                    41
                ],
                "Nature": 8,
                "Stats": [
                    124,
                    72,
                    35,
                    66,
                    65,
                    108
                ],
                "StatModifiers": [
                    0,
                    0,
                    0,
                    0,
                    0,
                    0,
                    0,
                    0,
                    0
                ],
                "StatusEffects": 0,
                "CurrentHP": 124,
                "HeldItem": {
                    "Id": 0,
                    "Name": "",
                    "Category": 0,
                    "FlingPower": 0,
                    "FlingEffect": 0,
                    "Flags": 0
                },
                "Moves": [
                    {
                        "Id": 433,
                        "CurrentPP": 0,
                        "MaxPP": 0,
                        "Name": "Trick Room",
                        "Type": 8192,
                        "Category": 0,
                        "Targets": 12,
                        "Priority": -7,
                        "Power": 0,
                        "Accuracy": 0,
                        "InitialMaxPP": 0
                    },
                    {
                        "Id": 225,
                        "CurrentPP": 0,
                        "MaxPP": 0,
                        "Name": "Dragon Breath",
                        "Type": 32768,
                        "Category": 2,
                        "Targets": 10,
                        "Priority": 0,
                        "Power": 60,
                        "Accuracy": 100,
                        "InitialMaxPP": 0
                    },
                    {
                        "Id": 232,
                        "CurrentPP": 0,
                        "MaxPP": 0,
                        "Name": "Metal Claw",
                        "Type": 256,
                        "Category": 1,
                        "Targets": 10,
                        "Priority": 0,
                        "Power": 50,
                        "Accuracy": 95,
                        "InitialMaxPP": 0
                    },
                    {
                        "Id": 59,
                        "CurrentPP": 0,
                        "MaxPP": 0,
                        "Name": "Blizzard",
                        "Type": 16384,
                        "Category": 2,
                        "Targets": 11,
                        "Priority": 0,
                        "Power": 110,
                        "Accuracy": 70,
                        "InitialMaxPP": 0
                    }
                ],
                "Friendship": 0,
                "OriginalTrainerID": 0,
                "Type": 4096
            }
        },
        {
            "Party": 0,
            "Slot": 0,
            "Team": 1,
            "Pokemon": {
                "Name": "Mew",
                "NatDex": 151,
                "Level": 6,
                "Ability": 108,
                "TotalExperience": 179,
                "Gender": 0,
                "IVs": [
                    3,
                    1,
                    1,
                    4,
                    0,
                    3
                ],
                "EVs": [
                    7,
                    24,
                    41,
                    11,
                    3,
                    20
                ],
                "Nature": 8,
                "Stats": [
                    28,
                    17,
                    17,
                    17,
                    17,
                    17
                ],
                "StatModifiers": [
                    0,
                    0,
                    0,
                    0,
                    0,
                    0,
                    0,
                    0,
                    0
                ],
                "StatusEffects": 0,
                "CurrentHP": 28,
                "HeldItem": {
                    "Id": 0,
                    "Name": "",
                    "Category": 0,
                    "FlingPower": 0,
                    "FlingEffect": 0,
                    "Flags": 0
                },
                "Moves": [
                    {
                        "Id": 61,
                        "CurrentPP": 0,
                        "MaxPP": 0,
                        "Name": "Bubble Beam",
                        "Type": 1024,
                        "Category": 2,
                        "Targets": 10,
                        "Priority": 0,
                        "Power": 65,
                        "Accuracy": 100,
                        "InitialMaxPP": 0
                    },
                    {
                        "Id": 457,
                        "CurrentPP": 0,
                        "MaxPP": 0,
                        "Name": "Head Smash",
                        "Type": 32,
                        "Category": 1,
                        "Targets": 10,
                        "Priority": 0,
                        "Power": 150,
                        "Accuracy": 80,
                        "InitialMaxPP": 0
                    },
                    {
                        "Id": 434,
                        "CurrentPP": 0,
                        "MaxPP": 0,
                        "Name": "Draco Meteor",
                        "Type": 32768,
                        "Category": 2,
                        "Targets": 10,
                        "Priority": 0,
                        "Power": 130,
                        "Accuracy": 90,
                        "InitialMaxPP": 0
                    },
                    {
                        "Id": 359,
                        "CurrentPP": 0,
                        "MaxPP": 0,
                        "Name": "Hammer Arm",
                        "Type": 2,
                        "Category": 1,
                        "Targets": 10,
                        "Priority": 0,
                        "Power": 100,
                        "Accuracy": 90,
                        "InitialMaxPP": 0
                    }
                ],
                "Friendship": 0,
                "OriginalTrainerID": 0,
                "Type": 8192
            }
        }
    ]
}
```
</details

### PR DESCRIPTION
**Pull request by elenirotsides**
When this is all said and done, it'll close #15 . Right now, I've commited something that I've written that I'm not very confident in, and would like some feedback on. I don't really know what this is supposed to look like, but here is what I have for now! Please let me know what I need to fix, what I've omitted, what I don't need, and what I'm doing wrong; I'm sure there's plenty. Also, I plopped some json data in a file called data.json for now, this is obviously not going to be needed once we start interacting with the api.﻿

### CODE REVIEW

File: pkmntypes.py [hunk 1]
```diff
@@ -22,3 +22,28 @@ def __init__(self, **kwargs):
...
+	"""
+	# will add make the docstrings comply with the sphinx format once this class is looking good
+
+	def __init__(self):
+		self.battle = battle
+		self.pokemon = pokemon
+		self.team = team
+		self.targets = targets
+		self.allies = allies
+		self.opponents = opponents
```
Comment (dyc3) [inline pkmntypes.py:38]: in this context, all the variables on the right side are not initialized

File: pkmntypes.py [hunk 2]
```diff
@@ -22,3 +22,28 @@ def __init__(self, **kwargs):
...
+	def parseJSON(self) -> dict:
+		with open('data.json') as d:
+			data = json.load(d)
+
+		battle = data['Battle']
+		pokemon = data['Pokemon']
+		team = data['Team']
+		targets = data['Targets']
+		allies = data['Allies']
+		opponents = data['Opponents']
```
Comment (dyc3) [inline pkmntypes.py:49]: these need to start with `self.` or it won't work

File: pkmntypes.py [hunk 3]
```diff
@@ -22,3 +22,28 @@ def __init__(self, **kwargs):
...
+
+	def __init__(self):
+		self.battle = battle
+		self.pokemon = pokemon
+		self.team = team
+		self.targets = targets
+		self.allies = allies
+		self.opponents = opponents
+
+	def parseJSON(self) -> dict:
```
Comment (dyc3) [inline pkmntypes.py:40]: Would rather have usage be like this:
```python
data = json.loads(text)
context = BattleContext(**data)
```

**Review by dyc3 — CHANGES_REQUESTED** [review]

File: pkmntypes.py (hunk 3 above)
Comment (elenirotsides) [reply pkmntypes.py:40]: Would `context` be initialized inside the `BattleContext` class?

Comment (dyc3) [reply pkmntypes.py:40]: No, `context` in this sample _is_ an instance of `BattleContext`. This is meant to be an example of how you would use the class in practice.

File: pkmntypes.py (hunk 1 above)
Comment (elenirotsides) [reply pkmntypes.py:38]: To fix this, would I need to set it equal to `data['key']` and then get rid of the other function I wrote below? @dyc3

Comment (dyc3) [reply pkmntypes.py:38]: yeah, basically

**Comment by elenirotsides** [comment]
Just force pushed up an edit, please let me know what you think ie if I'm offbase, missing something, etc! When you lmk that it looks good, then I'll add the sphinx docstring @dyc3

File: pkmntypes.py [hunk 4]
```diff
@@ -22,3 +22,20 @@ def __init__(self, **kwargs):
...
+
+class BattleContext():
+	"""This is a class that describes the point of view of a given Pokemon on the battlefield. It provides
+	enough information for a user to make an informed decision about what turn to make next.   
+	"""
+	# will add make the docstrings comply with the sphinx format once this class is looking good
+
+	def __init__(self):
+		with open('data.json') as d:
+			data = json.load(d)
```
Comment (dyc3) [inline pkmntypes.py:34]:
```suggestion
	def __init__(self, **kwargs):
```
and change `data` to `kwargs` below

**Review by dyc3 — CHANGES_REQUESTED** [review]
Looks good, but lets go ahead and remove the `data.json`

**Comment by elenirotsides** [comment]
@dyc3 should I also delete the `data.json` file as well?

**Comment by dyc3** [comment]
yes

File: pkmntypes.py [hunk 5]
```diff
@@ -22,3 +22,18 @@ def __init__(self, **kwargs):
...
+	# will add make the docstrings comply with the sphinx format once this class is looking good
+
+	def __init__(self, **kwargs):
+
+		self.battle = data['Battle']
+		self.pokemon = data['Pokemon']
+		self.team = data['Team']
+		self.targets = data['Targets']
+		self.allies = data['Allies']
+		self.opponents = data['Opponents']
```
Comment (dyc3) [inline pkmntypes.py:39]: you also need to change these as per my review comment

Comment (elenirotsides) [reply pkmntypes.py:39]: Yup, I already did, but I'm gonna push up those changes with the docstring edit. Unless you want me to commit them separately, lmk

Comment (dyc3) [reply pkmntypes.py:39]: No just push them up together

File: pkmntypes.py [hunk 6]
```diff
@@ -22,3 +22,30 @@ def __init__(self, **kwargs):
 
 class Pokemon():
 	pass
+
+class BattleContext():
+	"""This is a class that describes the point of view of a given Pokemon on the battlefield. It provides
+	enough information for a user to make an informed decision about what turn to make next. 
+
+	:param battle: The state of the battlefield
+	:type battle: dict
```
Comment (dyc3) [inline pkmntypes.py:31]: These type don't need to be specified up here...

File: pkmntypes.py [hunk 7]
```diff
@@ -22,3 +22,30 @@ def __init__(self, **kwargs):
...
+	:type targets: dict
+	:param allies: Ally targets in relation to the `Pokemon`
+	:type allies: dict
+	:param opponents: Enemy targets in relation to the `Pokemon`
+	:type opponents: dict
+	"""
+	
+	def __init__(self, **kwargs):
+
+		self.battle = kwargs['Battle']
```
Comment (dyc3) [inline pkmntypes.py:46]: ...because they should be specified down here.

eg. `self.battle: dict = kwargs['Battle']`

**Review by dyc3 — CHANGES_REQUESTED** [review]

File: pkmntypes.py (hunk 6 above)
Comment (elenirotsides) [reply pkmntypes.py:31]: I didn't really know what to do lol I'll delete them 😂  what about param, is that fine?

File: pkmntypes.py (hunk 7 above)
Comment (elenirotsides) [reply pkmntypes.py:46]: OHHHHHHHHHH

**Review by dyc3 — APPROVED** [review]

**Pull request merged**